Static AttributeKey in HttpPipelineBuilder causes native compilation to fail

In this PR for 4.0.0 https://github.com/micronaut-projects/micronaut-core/pull/6842 a static AttributeKey was added to the `HttpPipelineBuilder` class

This is causing Graal to fail with:

```
Caused by: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: No instances of io.netty.util.AttributeKey are allowed in the image heap as this class should be initialized at image runtime. Object has been initialized by the io.micronaut.http.server.netty.HttpPipelineBuilder class initializer with a trace:
        at io.netty.util.AttributeKey.<init>(AttributeKey.java:25)
        at io.netty.util.AttributeKey$1.newConstant(AttributeKey.java:30)
        at io.netty.util.AttributeKey$1.newConstant(AttributeKey.java:27)
        at io.netty.util.ConstantPool.createOrThrow(ConstantPool.java:101)
        at io.netty.util.ConstantPool.newInstance(ConstantPool.java:90)
        at io.netty.util.AttributeKey.newInstance(AttributeKey.java:55)
        at io.micronaut.http.server.netty.HttpPipelineBuilder.<clinit>(HttpPipelineBuilder.java:88)
```

Previously we fixed this like so: https://github.com/micronaut-projects/micronaut-core/pull/8013

Spotted as part of working on https://github.com/micronaut-projects/micronaut-core/issues/7691 to catch this sort of thing 😄 going forward on 4.0.0

--- a/http-server-netty/src/main/java/io/micronaut/http/server/netty/HttpPipelineBuilder.java
+++ b/http-server-netty/src/main/java/io/micronaut/http/server/netty/HttpPipelineBuilder.java
@@ -86,6 +86,7 @@ import java.util.function.Supplier;
  * @author ywkat
  */
 final class HttpPipelineBuilder {
+
     private static final Logger LOG = LoggerFactory.getLogger(HttpPipelineBuilder.class);
 
     private final NettyHttpServer server;
@@ -468,7 +469,7 @@ final class HttpPipelineBuilder {
          * and netty requests, and routing.
          */
         private void insertMicronautHandlers() {
-            channel.attr(getStreamPipelineAttribute().get()).set(this);
+            channel.attr(StreamPipelineAttributeKeyHolder.getInstance()).set(this);
 
             pipeline.addLast(ChannelPipelineCustomizer.HANDLER_HTTP_COMPRESSOR, new SmartHttpContentCompressor(embeddedServices.getHttpCompressionStrategy()));
             pipeline.addLast(ChannelPipelineCustomizer.HANDLER_HTTP_DECOMPRESSOR, new HttpContentDecompressor());
@@ -520,6 +521,22 @@ final class HttpPipelineBuilder {
                 }
                 pipeline.addLast(name, outboundHandlerAdapter);
             }
+        }
+    }
+
+    // We need the AttributeKey to be static, as it's used in NettyHttpRequest, but we can't eagerly initialize it
+    // as it would fail in Graal
+    static class StreamPipelineAttributeKeyHolder {
+
+        private StreamPipelineAttributeKeyHolder() {
+        }
+
+        private static class InstanceHolder {
+            private static final AttributeKey<StreamPipeline> INSTANCE = AttributeKey.newInstance("stream-pipeline");
+        }
+
+        static AttributeKey<StreamPipeline> getInstance() {
+            return InstanceHolder.INSTANCE;
         }
     }
 }

--- a/http-server-netty/src/main/java/io/micronaut/http/server/netty/HttpPipelineBuilder.java
+++ b/http-server-netty/src/main/java/io/micronaut/http/server/netty/HttpPipelineBuilder.java
@@ -521,19 +521,20 @@ final class HttpPipelineBuilder {
 
     // We need the AttributeKey to be static, as it's used in NettyHttpRequest, but we can't eagerly initialize it
     // as it would fail in Graal
-    static class StreamPipelineAttributeKeyHolder {
+    static final class StreamPipelineAttributeKeyHolder {
 
-        private static final AtomicReference<AttributeKey<StreamPipeline>> instance = new AtomicReference<>();
+        private static final AtomicReference<AttributeKey<StreamPipeline>> INSTANCE = new AtomicReference<>();
 
         private StreamPipelineAttributeKeyHolder() {
         }
 
         static AttributeKey<StreamPipeline> getInstance() {
-            return instance.updateAndGet(key -> {
+            return INSTANCE.updateAndGet(key -> {
                 if (key == null) {
                     return AttributeKey.newInstance("micronaut-stream-pipeline");
                 }
                 return key;
             });
         }
-    }}
+    }
+}

--- a/http-server-netty/src/main/java/io/micronaut/http/server/netty/HttpPipelineBuilder.java
+++ b/http-server-netty/src/main/java/io/micronaut/http/server/netty/HttpPipelineBuilder.java
@@ -74,6 +74,7 @@ import java.nio.channels.ClosedChannelException;
 import java.time.Duration;
 import java.time.Instant;
 import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Supplier;
 
 /**
@@ -528,15 +529,17 @@ final class HttpPipelineBuilder {
     // as it would fail in Graal
     static class StreamPipelineAttributeKeyHolder {
 
+        private static final AtomicReference<AttributeKey<StreamPipeline>> instance = new AtomicReference<>();
+
         private StreamPipelineAttributeKeyHolder() {
         }
 
-        private static class InstanceHolder {
-            private static final AttributeKey<StreamPipeline> INSTANCE = AttributeKey.newInstance("stream-pipeline");
-        }
-
         static AttributeKey<StreamPipeline> getInstance() {
-            return InstanceHolder.INSTANCE;
+            return instance.updateAndGet(key -> {
+                if (key == null) {
+                    return AttributeKey.newInstance("micronaut-stream-pipeline");
+                }
+                return key;
+            });
         }
-    }
-}
+    }}

--- a/http-server-netty/src/main/java/io/micronaut/http/server/netty/HttpPipelineBuilder.java
+++ b/http-server-netty/src/main/java/io/micronaut/http/server/netty/HttpPipelineBuilder.java
@@ -74,6 +74,7 @@ import java.nio.channels.ClosedChannelException;
 import java.time.Duration;
 import java.time.Instant;
 import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.Supplier;
 
 /**
  * Helper class that manages the {@link ChannelPipeline} of incoming HTTP connections.
@@ -85,8 +86,6 @@ import java.util.concurrent.ThreadLocalRandom;
  * @author ywkat
  */
 final class HttpPipelineBuilder {
-    static final AttributeKey<StreamPipeline> STREAM_PIPELINE_ATTRIBUTE = AttributeKey.newInstance("stream-pipeline");
-
     private static final Logger LOG = LoggerFactory.getLogger(HttpPipelineBuilder.class);
 
     private final NettyHttpServer server;
@@ -135,6 +134,11 @@ final class HttpPipelineBuilder {
 
     boolean supportsSsl() {
         return sslContext != null;
+    }
+
+    static Supplier<AttributeKey<StreamPipeline>> getStreamPipelineAttribute() {
+        AttributeKey<StreamPipeline> attributeKey = AttributeKey.newInstance("stream-pipeline");
+        return () -> attributeKey;
     }
 
     final class ConnectionPipeline {
@@ -464,7 +468,7 @@ final class HttpPipelineBuilder {
          * and netty requests, and routing.
          */
         private void insertMicronautHandlers() {
-            channel.attr(STREAM_PIPELINE_ATTRIBUTE).set(this);
+            channel.attr(getStreamPipelineAttribute().get()).set(this);
 
             pipeline.addLast(ChannelPipelineCustomizer.HANDLER_HTTP_COMPRESSOR, new SmartHttpContentCompressor(embeddedServices.getHttpCompressionStrategy()));
             pipeline.addLast(ChannelPipelineCustomizer.HANDLER_HTTP_DECOMPRESSOR, new HttpContentDecompressor());

--- a/http-server-netty/src/main/java/io/micronaut/http/server/netty/HttpPipelineBuilder.java
+++ b/http-server-netty/src/main/java/io/micronaut/http/server/netty/HttpPipelineBuilder.java
@@ -75,7 +75,6 @@ import java.time.Duration;
 import java.time.Instant;
 import java.util.concurrent.ThreadLocalRandom;
 import java.util.concurrent.atomic.AtomicReference;
-import java.util.function.Supplier;
 
 /**
  * Helper class that manages the {@link ChannelPipeline} of incoming HTTP connections.
@@ -136,11 +135,6 @@ final class HttpPipelineBuilder {
 
     boolean supportsSsl() {
         return sslContext != null;
-    }
-
-    static Supplier<AttributeKey<StreamPipeline>> getStreamPipelineAttribute() {
-        AttributeKey<StreamPipeline> attributeKey = AttributeKey.newInstance("stream-pipeline");
-        return () -> attributeKey;
     }
 
     final class ConnectionPipeline {

--- a/http-server-netty/src/main/java/io/micronaut/http/server/netty/NettyHttpRequest.java
+++ b/http-server-netty/src/main/java/io/micronaut/http/server/netty/NettyHttpRequest.java
@@ -527,7 +527,7 @@ public class NettyHttpRequest<T> extends AbstractNettyHttpRequest<T> implements 
             );
 
             int ourStream = ((Http2StreamChannel) channelHandlerContext.channel()).stream().id();
-            HttpPipelineBuilder.StreamPipeline originalStreamPipeline = channelHandlerContext.channel().attr(HttpPipelineBuilder.STREAM_PIPELINE_ATTRIBUTE).get();
+            HttpPipelineBuilder.StreamPipeline originalStreamPipeline = channelHandlerContext.channel().attr(HttpPipelineBuilder.getStreamPipelineAttribute().get()).get();
 
             new Http2StreamChannelBootstrap(channelHandlerContext.channel().parent())
                     .handler(new ChannelInitializer<Http2StreamChannel>() {

--- a/http-server-netty/src/main/java/io/micronaut/http/server/netty/NettyHttpRequest.java
+++ b/http-server-netty/src/main/java/io/micronaut/http/server/netty/NettyHttpRequest.java
@@ -527,7 +527,7 @@ public class NettyHttpRequest<T> extends AbstractNettyHttpRequest<T> implements 
             );
 
             int ourStream = ((Http2StreamChannel) channelHandlerContext.channel()).stream().id();
-            HttpPipelineBuilder.StreamPipeline originalStreamPipeline = channelHandlerContext.channel().attr(HttpPipelineBuilder.getStreamPipelineAttribute().get()).get();
+            HttpPipelineBuilder.StreamPipeline originalStreamPipeline = channelHandlerContext.channel().attr(HttpPipelineBuilder.StreamPipelineAttributeKeyHolder.getInstance()).get();
 
             new Http2StreamChannelBootstrap(channelHandlerContext.channel().parent())
                     .handler(new ChannelInitializer<Http2StreamChannel>() {

